### ISSUE TITLE
Support mayInterruptIfRunning flag in CompletableFuture.await extension

### ISSUE THREAD

**Issue by elizarov** [Kotlin/kotlinx.coroutines#259]
This is useful for cases when we are working with interruptible blocking APIs. See discussion in #79

**Comment by qwwdfsad** [comment]
It cannot be implemented: `CompletableFuture` doesn't use flag `mayInterruptIfRunning` and classic `Future` is not a `CompletionStage` and is not awaitable

**Comment by qwwdfsad** [comment]
Citing https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/CompletableFuture.html here:
```
mayInterruptIfRunning - this value has no effect in this implementation because interrupts are not used to control processing.

```

**Comment by revintec** [comment]
@qwwdfsad thanks for the link and I should have provided more details:

1. quoting your doc, this value has no effect in `this implementation` -- emphasis mine
I don't think we should design APIs based (only) on `current implementation`s. as this method is non-final, subclass may very well using this flag for finer controls

2. I did find a subclass in a very common code path -- awaiting HttpClient's async exchanges
a `CompletableFuture` of type `MinimalFuture` returned from here `https://github.com/openjdk/jdk17/blob/74007890bb9a3fa3a65683a3f480e399f2b1a0b6/src/java.net.http/share/classes/jdk/internal/net/http/HttpClientImpl.java#L653`

the corresponding coroutine waits on the resulting future using `httpClient.sendAsync(...).await()`, and subsequently get cancelled(due to timeout, etc). but the underlying http request can't be cancelled unless the `mayInterruptIfRunning` flag is correctly passed through -- as value of `true`, see
https://github.com/openjdk/jdk17/blob/74007890bb9a3fa3a65683a3f480e399f2b1a0b6/src/java.net.http/share/classes/jdk/internal/net/http/common/MinimalFuture.java#L107
https://github.com/openjdk/jdk17/blob/74007890bb9a3fa3a65683a3f480e399f2b1a0b6/src/java.net.http/share/classes/jdk/internal/net/http/MultiExchange.java#L276

**Comment by qwwdfsad** [comment]
@revintec Thanks!

Indeed there probably **exist** some implementations that use the flag, but this fact alone is not enough to expand our API surface. For now, it's not even clear whether these implementations from HttpClient leak to user code at all.

What we need is actual use cases where it is clear which exact **user-facing** problem we are solving with the proposed API addition

**Comment by dkhalanskyjb** [comment]
I'd think `true` would be a more idiomatic value to use here, so maybe we should also consider the alternative of just always passing `true`. When we `cancel` in coroutines, we almost always intend to cancel the computation if it was already started.

**Comment by fvasco** [comment]
Some hours ago I used `HttpClient` and looked for an `awaitInterruptible` method, similar to `runInterruptible`.

**Comment by dkhalanskyjb** [comment]
The problem is, is there anything in the `HttpClient` documentation that indicates that cancelling the future will interrupt the process? https://docs.oracle.com/en/java/javase/12/docs/api/java.net.http/java/net/http/HttpClient.html I don't see anything in that vein. So, should one rely on this behavior?

**Comment by fvasco** [comment]
@dkhalanskyjb is it now possible to reopen this issue?

**Comment by dkhalanskyjb** [comment]
Yep, you're right: https://docs.oracle.com/en/java/javase/16/docs/api//java.net.http/java/net/http/HttpClient.html#sendAsync(java.net.http.HttpRequest,java.net.http.HttpResponse.BodyHandler,java.net.http.HttpResponse.PushPromiseHandler) So, yes, this gives us a use case: cancelling `HttpClient.sendAsync` on JDK 16.

Now, a question is, could someone want to parameterize this, or is `cancel(true)` always the right choice? Boolean flags are quite inconvenient, as is having two `await` functions that behave exactly the same for 99% (?) of `CompletableFuture`s out there. If someone intentionally uses our version with `cancel(false)`, of course, there are no other options.

**Comment by fvasco** [comment]
Similar API with the same issue:

* `CompletionStage<T>.asDeferred()`
* `CancellableContinuation<*>.cancelFutureOnCancellation(future: Future<*>)`
